### ISSUE TITLE
iot connection error for Authenticated role

### ISSUE THREAD

**Issue by aherpooja** [aws-amplify/aws-sdk-ios#3266]
**Error occurred while connecting to iot using authenticated user Cognito pool**

**Services:**
- AWSCognitoIdentityUserPool
- AWSIoT
- AWSIoTDataManager

**User cognito pool, lambda, dynamo db is on USEast2 region and IOT devices is on USEast1 region**

@objc var iotDataManager: AWSIoTDataManager!
@objc var iot: AWSIoT!

//MARK:- Initialization
    func initializeAWS(completionHandler : @escaping (Bool)->Void) {
        initializeControlPlane()
        completionHandler(true)
    }
    
    func initializeControlPlane() {
        initializeControlPlane(credentialsProvider: AWSMobileClient.default())
        initializeDataPlane(credentialsProvider: AWSMobileClient.default())
        
        let attachPolicyRequest = AWSIoTAttachPolicyRequest()
        attachPolicyRequest!.policyName = AWSConstants.PolicyName
        
        AWSMobileClient.default().getIdentityId().continueWith(block: { (task) -> AnyObject? in
            if (task.error != nil) {
                Logger.log("Error: " + task.error!.localizedDescription)
            }
            else {
                // the task result will contain the identity id
                let cognitoId = task.result!
                attachPolicyRequest!.target = cognitoId as String
                Logger.log("Cognito id: \(cognitoId)")
                self.iot.attachPolicy(attachPolicyRequest!) { (error) in
                    if error != nil {
                        Logger.log("error attach policy \(error!)")
                    }
                }
            }
            return task
        })
    }
    
    func initializeControlPlane(credentialsProvider: AWSCredentialsProvider) {
        //Initialize control plane
        // Initialize the Amazon Cognito credentials provider
        let controlPlaneServiceConfiguration = AWSServiceConfiguration(region:AWSConstants.AWSIoTRegion, credentialsProvider:credentialsProvider)
        
        //IoT control plane seem to operate on iot.<region>.amazonaws.com
        //Set the defaultServiceConfiguration so that when we call AWSIoTManager.default(), it will get picked up
        AWSServiceManager.default().defaultServiceConfiguration = controlPlaneServiceConfiguration
        iot = AWSIoT.default()
    }
    
    func initializeDataPlane(credentialsProvider: AWSCredentialsProvider) {
        //Initialize Dataplane:
        // IoT Dataplane must use your account specific IoT endpoint
        let iotEndPoint = AWSEndpoint(urlString: AWSConstants.IOTEndPoint)
        
        let iotDataConfiguration = AWSServiceConfiguration(region: AWSConstants.AWSIoTRegion,
                                                           endpoint: iotEndPoint,
                                                           credentialsProvider: credentialsProvider)
        
        //IoTData manager operates on xxxxxxx-iot.<region>.amazonaws.com
        AWSIoTDataManager.register(with: iotDataConfiguration!, forKey: AWSConstants.ASWIoTDataManager)
        iotDataManager = AWSIoTDataManager(forKey: AWSConstants.ASWIoTDataManager)
        
        connectAWSViaWebsocket { (_) in
            
        }
    }
    
    func connectAWSViaWebsocket(completionHandler : @escaping (Bool)->Void) {
        func mqttEventCallbackWebsocket(_ status: AWSIoTMQTTStatus) {
            switch status {
            case .connecting:
                Logger.log("Connecting to AWS IoT")
                deviceConnectionStatus = .connecting
                
            case .connected:
                deviceConnectionStatus = .connected
                
                Logger.log("Connected to AWS successfully")
                self.subscribe() {deviceStauts in
                    completionHandler(deviceStauts)
                }
            case .connectionError:
                Logger.log("AWS IoT connection error")
                deviceConnectionStatus = .connectionError
                
            case .connectionRefused:
                Logger.log("AWS IoT connection refused")
                deviceConnectionStatus = .connectionError
                
            case .protocolError:
                Logger.log("AWS IoT protocol error")
                deviceConnectionStatus = .connectionError
                
            case .disconnected:
                Logger.log("AWS disconnected")
                deviceConnectionStatus = .disconnected
                
            case .unknown:
                Logger.log("AWS unknown connection")
                deviceConnectionStatus = .connectionError
                
            default:
                Logger.log("AWS IoT unknown state")
                deviceConnectionStatus = .connectionError
            }
            DispatchQueue.main.async {
                NotificationCenter.default.post( name: Constants.connectionStatusChanged, object: self )
            }
        }
        
        let uuid = UUID().uuidString
        iotDataManager.connectUsingWebSocket(withClientId: uuid, cleanSession: true, statusCallback: mqttEventCallbackWebsocket(_:))
    }

**Environment(please complete the following information):**
 - SDK Version: 2.19.1
 - Dependency Manager: Cocoapods
 - Swift Version : 5.0

**Device Information (please complete the following information):**
 - Device: All
 - iOS Version: all
 - Specific to simulators: Issue on both Device and simulator

**Xcode 12.2 Log:**
2020-11-26 16:32:30:238 PrimoAger[11276:286763] Response body:
{"message":"Policy not found"}
2020-11-26 16:32:30 [AWSIoTService.swift initializeControlPlane() #68]: error attach policy Error Domain=com.amazonaws.AWSIoTErrorDomain Code=19 "Policy not found" UserInfo={NSLocalizedDescription=Policy not found, NSLocalizedFailureReason=ResourceNotFoundException}
2020-11-26 16:32:30:394 PrimoAger[11276:286763] Websocket did open and is connected.
2020-11-26 16:32:30:394 PrimoAger[11276:286837] <<<NSThread: 0x6000031f5400>{number = 11, name = (null)}>> Initializing MQTTEncoder and MQTTDecoder streams
2020-11-26 16:32:30:395 PrimoAger[11276:286837] opening encoder stream.
2020-11-26 16:32:30:395 PrimoAger[11276:286837] opening decoder stream.
2020-11-26 16:32:30 [AWSIoTService.swift mqttEventCallbackWebsocket(_:) #116]: AWS IoT connection error
2020-11-26 16:32:30:679 PrimoAger[11276:286763] WebSocket closed with code:1001 with reason:Stream end encountered
2020-11-26 16:32:30:679 PrimoAger[11276:286837] MQTT session error, code: 2
2020-11-26 16:32:30:682 PrimoAger[11276:286837] closing encoder stream.
2020-11-26 16:32:30:682 PrimoAger[11276:286837] closing decoder stream.
2020-11-26 16:32:31:686 PrimoAger[11276:286840] Attempting to reconnect.
2020-11-26 16:32:31:686 PrimoAger[11276:286840] AWSIoTMQTTClient: connecting via websocket. 
2020-11-26 16:32:31:686 PrimoAger[11276:286840] Using AWSCredentials.
2020-11-26 16:32:31 [AWSIoTService.swift mqttEventCallbackWebsocket(_:) #116]: AWS IoT connection error
2020-11-26 16:32:31:687 PrimoAger[11276:286837] MQTTSessionEventConnectionClosed: MQTT session closed.
2020-11-26 16:32:31:688 PrimoAger[11276:286766] user metadata is : ?SDK=iOS&Version=2.19.1
2020-11-26 16:32:31 [AWSIoTService.swift mqttEventCallbackWebsocket(_:) #105]: Connecting to AWS IoT
2020-11-26 16:32:31:688 PrimoAger[11276:286766] Metrics collection is: Enabled
2020-11-26 16:32:32:878 PrimoAger[11276:286763] Websocket did open and is connected.
2020-11-26 16:32:32:878 PrimoAger[11276:286906] <<<NSThread: 0x6000031f6fc0>{number = 14, name = (null)}>> Initializing MQTTEncoder and MQTTDecoder streams
2020-11-26 16:32:32:878 PrimoAger[11276:286906] opening encoder stream.
2020-11-26 16:32:32:878 PrimoAger[11276:286906] opening decoder stream.
2020-11-26 16:32:33:114 PrimoAger[11276:286766] WebSocket closed with code:1001 with reason:Stream end encountered
2020-11-26 16:32:33 [AWSIoTService.swift mqttEventCallbackWebsocket(_:) #116]: AWS IoT connection error

**Comment by wooj2** [comment]
Hi @aherpooja ,

Based on the logs, it looks like you are attempting to assume an authenticated role while accessing an IoT Core endpoint over a websocket, is this correct?  If not, can you please tell us what your use case is?

Have you double checked the IAM policies that are attached to the role you are attempting to assume?  Can you please double check the policy document that is stored in IoT Core?  Has this worked previously for you, or has something changed?

Thanks!

**Comment by aherpooja** [comment]
Hi, @wooj2 Previously unauthenticated users were also allowed to access the IoT. But now they have removed unauthenticated role. I have resolved the connection issue. It seems I was setting the wrong region for IoT. After debugging the code, I have found that below code sets the region set from the configuration file which was the wrong region for IoT
iot = AWSIoT.default()


Thank you for looking into the issue
